edb/edgeql/compiler/stmt.py

```diff
@@ -362,6 +333,22 @@ def compile_InsertQuery(
             ctx=ctx,
         )
 
+        if expr.on_conflict is not None:
+            constraint_spec, else_branch = expr.on_conflict
+
+            if constraint_spec:
+                with ictx.new() as constraint_ctx:
+                    stmt.on_conflict = compile_insert_on_conflict_constraint(
+                        stmt, constraint_spec, ctx=constraint_ctx)
+            else:
+                stmt.on_conflict = True
+
+            if else_branch:
+                raise errors.QueryError(
```

Comment:
nit: an `UnsupportedFeatureError` error would be more appropriate. See the full list here: https://github.com/edgedb/edgedb/blob/master/edb/api/errors.txt#L26